File: scraper_service/services/pgUpdateHandler.py

""" 
操作整個DB的所有table的schema變化，而非專屬於scrape
由於爬蟲伺服器寫入較多，因此讓其管理整個DB
但其管理交由alembic，創建專案所需要的所有資料表
雖然缺少ORM操作的便利性，但比較易懂，容易直接測試SQL
但考慮交給migration紀錄變更
"""
import os
import psycopg2

# import json
from dotenv import load_dotenv
load_dotenv()

class PostgresHandler:
    def __init__(self):
        self.conn = psycopg2.connect(
            host=os.getenv("DB_HOST"),
            database=os.getenv("DB_NAME"),
            user=os.getenv("DB_USER"),
            password=os.getenv("DB_PASSWORD")
        )
        self.cur = self.conn.cursor()
    """
    jobs只先在參數驗證是list(在腳本層已經由JobModel來驗證)
    把dict value的驗證交由資料庫層
    """
    def insert_jobs_into_postgres(self, jobs: list[dict]):
        try:
            with open("../postgres_db/insert_jobs.sql", 'r', encoding='utf-8') as file:
                insert_jobs_sql = file.read()
            with self.conn.cursor() as cur:
                for job in jobs:
                    values = tuple(job.values())
                    cur.execute(insert_jobs_sql, values)
                self.conn.commit()
            print(f'成功將 {len(jobs)} 個職位信息插入到 PostgreSQL 表格中')
        except Exception as e:
          print(f'插入數據庫時出錯: {e}')
          self.conn.rollback()

File: scraper_service_js/services/pgUpdateHandler.js
const path = require('path');
const fs = require('fs');
const { Client } = require('pg');
require('dotenv').config();

const logger = require('../winston');

const createClient = () => {
  const client = new Client({
    host: process.env.DB_HOST,
    database: process.env.DB_NAME,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
  });

  return client;
};

const connectClient = async (client) => {
  try {
    await client.connect();
    logger.info('Connected to PostgreSQL', { source: 'db' });
    await createTables(client);
  } catch (err) {
    logger.error('Connection error', err.stack);
  }
};

const createTable = async (client, filePath) => {
  const sql = fs.readFileSync(path.resolve(__dirname, filePath), 'utf8');
  try {
    await client.query(sql);
    logger.info(`Table created or already exists from ${filePath}`, { source: 'db' });
  } catch (err) {
    logger.error(`Error creating table from ${filePath}:`, err);
  }
};

const createTriggerAndLog = async (client) => {
  const createTriggerLogSQL = fs.readFileSync(path.resolve(__dirname, '../../postgres_db/create_trigger_log.sql'), 'utf8');
  const createTriggerSQL = fs.readFileSync(path.resolve(__dirname, '../../postgres_db/create_act_pub_trigger.sql'), 'utf8');
  try {
    await client.query(createTriggerLogSQL);
    logger.info('Trigger log table created or already exists.', { source: 'db' });
    await client.query(createTriggerSQL);
    logger.info('Activation publication trigger created or already exists.', { source: 'db' });
  } catch (err) {
    logger.error('Error creating triggers:', err);
  }
};

const createTables = async (client) => {
  try {
    await createTable(client, '../../postgres_db/create_jobs.sql');
    await createTable(client, '../../postgres_db/create_job_subs.sql');
    await createTable(client, '../../postgres_db/create_id_subs.sql');
    await createTable(client, '../../postgres_db/create_act_pub_channel.sql');
    await createTriggerAndLog(client);
    logger.info('All tables and triggers created successfully.', { source: 'db' });
  } catch (err) {
    logger.error('Error creating tables:', err);
  }
};

const insertJobsIntoPostgres = async (client, jobs) => {
  const insertJobsSQL = fs.readFileSync(path.resolve(__dirname, '../../postgres_db/insert_jobs_js.sql'), 'utf8');
  try {
    for (const job of jobs) {
      await client.query(insertJobsSQL, [
        job.title,
        job.company_name,
        job.industry,
        job.experience,
        job.description,
        JSON.stringify(job.requirements),
        job.additional_conditions,
        job.salary,
        job.people,
        job.location,
        job.update_date,
        job.record_time,
        job.source,
        job.keywords,
        job.url,
      ]);
    }
    logger.info(`Successfully inserted ${jobs.length} job(s) into PostgreSQL.`, { source: 'db' });
  } catch (err) {
    logger.error('Error inserting data into PostgreSQL:', err);
  }
};

module.exports = { createClient, connectClient, insertJobsIntoPostgres };
